accepted/2020/feature-switch.md

```diff
@@ -0,0 +1,119 @@
+
+Feature switches for large areas of functionality which may span multiple libraries should be exposed as public APIs to make it easy and consistent. For example if there would be a feature switch to disable support for HTTP2 there should be a public API property indicating the desired behavior. This is because it's likely that even libraries outside of core libraries may need to change their behavior based on the property. Having one property in this case would enable linker to trim branches across multiple libraries in a consistent way.
+
+Feature switches for localized functionality which is not likely to span multiple libraries can use only internal properties.
+
+It's up to each feature what is the behavior when it's turned off. In lot of cases the code would probably throw an exception, but in some cases it may be able to fallback to different implementation. For example if HTTP2 is disabled the code may be able to use HTTP1 transparently.
+
+*Implementation note: The property which is used to determine if the feature is on or off should be written and used in such a way that tiering optimizations would be possible and JIT could trim away the "unused" code. Pattern which should achieve this in most cases is to get a read-only static bool property.*
+
+### Unified pattern for feature definition in MSBuild
+Introduce a standard pattern how to define these properties in the SDK. Basically a way to define a property and have it automatically passed through to `.runtimeconfig.json` as well as linker substitutions.
+
+This requires a mapping between the MSBuild property name, the full name of the feature switch for runtime configuration (which would show up in `.runtimeconfig.json`) and the read-only static property in the managed code which is used to branch the behavior. 
+
+*Note that it seems likely that some of switches may require 1:many mapping because they're targeting existing code which uses multiple properties to determine the presence of the feature.*
+
+Ultimately some of these switches may also need VS UI integration, but for now most of the feature switches should be perfectly fine with only MSBuild properties.
+
+The ability to specify these from MSBuild projects as properties also means that different templates and SDKs can set different defaults - so for example Blazor or Xamarin SDKs may choose to turn off some of these features by default.
+
+### Generate the right input for the linker in SDK
+Currently there's no integration of the substitutions feature in the linker with the `ILLink` task/targets. Introduce a standard way to  define the content of `substitutions.xml` from the SDK (MSBuild items/properties) and then implement the file generation in the `ILLink` task and pass it to the linker.
```

Comment:
We could have a separate SDK that supports manipulating IL, and all the things. There's one already but it's features are narrow!! We could expand the feature set of `Microsoft.NET.Sdk.IL` to include the capabilities that are discussed here?!